src/Java.Interop.Tools.JavaTypeSystem/Adapters/JavaXmlApiImporter.cs

```diff
@@ -0,0 +1,370 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+using Java.Interop.Tools.JavaTypeSystem.Models;
+
+namespace Java.Interop.Tools.JavaTypeSystem
+{
+	public class JavaXmlApiImporter
+	{
+		public static JavaTypeCollection ParseString (string xml, JavaTypeCollection? collection = null)
+		{
+			var doc = XDocument.Parse (xml, LoadOptions.SetBaseUri | LoadOptions.SetLineInfo);
+
+			return Parse (doc, collection);
+		}
+
+		public static JavaTypeCollection Parse (TextReader reader, JavaTypeCollection? collection = null)
+		{
+			var doc = XDocument.Load (reader, LoadOptions.SetBaseUri | LoadOptions.SetLineInfo);
+
+			return Parse (doc, collection);
+		}
+
+		public static JavaTypeCollection Parse (string filename, JavaTypeCollection? collection = null)
+		{
+			var doc = XDocument.Load (filename, LoadOptions.SetBaseUri | LoadOptions.SetLineInfo);
+
+			return Parse (doc, collection);
+		}
+
+		static JavaTypeCollection Parse (XDocument doc, JavaTypeCollection? collection = null)
+		{
+			collection ??= new JavaTypeCollection ();
+
+			var root = doc.Root;
+
+			if (root is null)
+				throw new Exception ("Invalid XML file");
+
+			collection.ApiSource = root.XGetAttributeOrNull ("api-source");
+			collection.Platform = root.XGetAttributeOrNull ("platform");
+
+			var packages = new List<JavaPackage> ();
+
+			foreach (var elem in root.Elements ()) {
+				switch (elem.Name.LocalName) {
+					case "package":
+						packages.Add (ParsePackage (elem, collection));
+						break;
+				}
+			}
+
+			// First add all non-nested types
+			foreach (var type in packages.SelectMany (p => p.Types).Where (t => !t.NestedName.Contains ('.')))
```

Comment:
This "all non-nested types followed by all nested types" feels "weird", in that it'll potentially separate nested types from their containing type.

For example, there is one `android.accessibilityservice.AccessibilityButtonController` nested type, `AccessibilityButtonController.AccessibilityButtonCallback`.

If we "split out" nested types in this fashion, we'd wind up with:

```xml
<class name="AccessibilityButtonController" …/>
<class name="AccessibilityGestureEvent" … />
<class name="AccessibilityService" … />
<!-- at least three other types… -->
<class name="AccessibilityButtonController.AccessibilityButtonCallback" … />
```

This differs from current output -- which is probably fine? -- but means that there's no "proximity" between nested types and their declaring types in document order.

This shouldn't break any tooling, but feels like something that could be "weird" or "annoying" for *human* readers.

Alternatively, could we *sort* `packages` in the order that we want?

Reply:
The output is done the way you suggest.  We output a type and then recurse into its nested types, so the nested types will always be directly after their declaring type.

https://github.com/xamarin/java.interop/pull/849/files#diff-41afcd0fb56d03aabb52fc26b23822c138457f30e40940ae9d8b43e018e02690R58-R67